huntsman/camera/pyro.py

```diff
@@ -0,0 +1,502 @@
+    def _clean_directories(self, source):
+        """
+        Clean up empty directories left behind by rsysc.
+        """
+        user_at_host, path = source.split(':')
+        path_root = path.split('/./')[0]
+        try:
+            result = subprocess.run(['ssh',
+                                     user_at_host,
+                                     'find {} -empty -delete'.format(path_root)],
+                                    check=True)
+        except subprocess.CalledProcessError as err:
+            msg = "Clean up of empty directories in {}:{} failed".format(user_at_host, path_root)
+            warn(msg)
+            self.logger.error(msg)
+            raise err
+        self.logger.debug("Clean up of empty directories in {}:{} complete".format(user_at_host,
+                                                                                   path_root))
+        return source
+
+    def _file_transfer(self, source, destination):
+        """
+        Used rsync to move a file from source to destination.
+        """
+        # Need to make sure the destination directory already exists because rsync isn't
+        # very good at creating directories.
+        os.makedirs(os.path.dirname(destination), mode=0o775, exist_ok=True)
+        try:
+            result = subprocess.run(['rsync',
+                                     '--archive',
+                                     '--relative',
+                                     '--recursive',
+                                     '--remove-source-files',
+                                     source,
+                                     destination],
+                                    check=True)
+        except subprocess.CalledProcessError as err:
+            msg = "File transfer {} -> {} failed".format(source, destination)
+            warn(msg)
+            self.logger.error(msg)
+            raise err
+        self.logger.debug("File transfer {} -> {} complete".format(source.split('/./')[1],
+                                                                   destination))
+        return source
+
+    def _async_wait(self, future_result, name='?', event=None, timeout=None):
+        # For now not checking for any problems, just wait for everything to return (or timeout)
+        if future_result.wait(timeout):
+            result = future_result.value
+        else:
+            msg = "Timeout while waiting for {} on {}".format(name, self.name)
+            warn(msg)
+            self.logger.error(msg)
+            return False
+
+        if event is not None:
+            event.set()
+
+        return result
+
+
+@Pyro4.expose
+@Pyro4.behavior(instance_mode="single")
+class CameraServer(object):
+    """
+    Wrapper for the camera class for use as a Pyro camera server
+    """
+    def __init__(self):
+        # Pyro classes ideally have no arguments for the constructor. Do it all from config file.
+        self.config = load_config(config_files=['pyro_camera.yaml'])
```

Comment:
With Pyro the recommended approach is to pass the class to the daemon rather than an object, and let the daemon handle object creation. That precludes passing any arguments to it, so I didn't even bother putting any in the constructor here. It is possible to pass an object instead though, which would enable overriding the config file name.